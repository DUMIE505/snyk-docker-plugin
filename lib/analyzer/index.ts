import * as osReleaseDetector from './os-release-detector';
import * as imageInspector from './image-inspector';
import * as apkAnalyzer from './apk-analyzer';
import * as aptAnalyzer from './apt-analyzer';
import * as rpmAnalyzer from './rpm-analyzer';
import * as binariesAnalyzer from './binaries-analyzer';
const debug = require('debug')('snyk');

export {
  analyze,
};

async function analyze(targetImage: string, options?: any) {
  const [
    imageInspection,
    osRelease,
    results,
  ] = await Promise.all([
    imageInspector.detect(targetImage, options),
    osReleaseDetector.detect(targetImage, options),
    Promise.all([
      apkAnalyzer.analyze(targetImage, options),
      aptAnalyzer.analyze(targetImage, options),
      rpmAnalyzer.analyze(targetImage, options),
    ]).catch((err) => {
      debug(`Error while running analyzer: '${err}'`);
      throw new Error(`Failed to detect installed OS packages: '${err}'`);
    }),

  ]);

  const { installedPackages, pkgManager } =
    getInstalledPackages(results as any[]);
  let binaries;
  try {
    binaries = await binariesAnalyzer.analyze(
      targetImage, installedPackages, pkgManager, options);
  } catch (err) {
    debug(`Error while running binaries analyzer: '${err}'`);
    throw new Error('Failed to detect binaries versions');
  }

  return {
    imageId: imageInspection.Id,
    osRelease,
    results,
    binaries,
    imageLayers: imageInspection.RootFS && imageInspection.RootFS.Layers,
  };
}

function getInstalledPackages(results: any[]):
 {installedPackages: string[], pkgManager?: string} {
  const dockerAnalysis = results.find((res) => {
    return res.Analysis && res.Analysis.length > 0;
  });

  if (!dockerAnalysis) {
    return { installedPackages: [] };
  }
  const installedPackages = dockerAnalysis.Analysis.map((pkg) => pkg.Name);
  let pkgManager = dockerAnalysis.AnalyzeType;
  if (pkgManager) {
    pkgManager = pkgManager.toLowerCase();
  }
  return { installedPackages, pkgManager };
}
