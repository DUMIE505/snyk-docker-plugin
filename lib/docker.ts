import { fileSync } from "tmp";
import {
  ExtractedImage,
  ExtractedKeyFiles,
  extractImageKeyFiles,
} from "./analyzer/image-extractor";
import { execute } from "./sub-process";

export { Docker, DockerOptions };

interface DockerOptions {
  host?: string;
  tlsVerify?: string;
  tlsCert?: string;
  tlsCaCert?: string;
  tlsKey?: string;
}

type SaveImageCallback = (err: any, name: string) => void;

class Docker {
  public static run(args: string[], options?: DockerOptions) {
    return execute("docker", [...Docker.createOptionsList(options), ...args]);
  }

  private static createOptionsList(options: any) {
    const opts: string[] = [];
    if (!options) {
      return opts;
    }
    if (options.host) {
      opts.push(`--host=${options.host}`);
    }
    if (options.tlscert) {
      opts.push(`--tlscert=${options.tlscert}`);
    }
    if (options.tlscacert) {
      opts.push(`--tlscacert=${options.tlscacert}`);
    }
    if (options.tlskey) {
      opts.push(`--tlskey=${options.tlskey}`);
    }
    if (options.tlsverify) {
      opts.push(`--tlsverify=${options.tlsverify}`);
    }
    return opts;
  }

  private optionsList: string[];

  constructor(private targetImage: string, options?: DockerOptions) {
    this.optionsList = Docker.createOptionsList(options);
  }

  public run(cmd: string, args: string[] = []) {
    return execute("docker", [
      ...this.optionsList,
      "run",
      "--rm",
      "--entrypoint",
      '""',
      "--network",
      "none",
      this.targetImage,
      cmd,
      ...args,
    ]);
  }

  public async inspect(targetImage: string) {
    return await execute("docker", [
      ...this.optionsList,
      "inspect",
      targetImage,
    ]);
  }

  public async catSafe(filename: string) {
    try {
      return await this.run("cat", [filename]);
    } catch (error) {
      const stderr: string = error.stderr;
      if (typeof stderr === "string") {
        if (
          stderr.indexOf("No such file") >= 0 ||
          stderr.indexOf("file not found") >= 0
        ) {
          return { stdout: "", stderr: "" };
        }
      }
      throw error;
    }
  }

  public async save(callback: SaveImageCallback): Promise<any> {
    const tmpobj = fileSync({
      mode: 0o644,
      prefix: "docker-",
      postfix: ".image",
      detachDescriptor: true,
    });
    let err = "";

    try {
      await execute("docker", [
        ...this.optionsList,
        "save",
        "-o",
        tmpobj.name,
        this.targetImage,
      ]);
    } catch (error) {
      const stderr: string = error.stderr;
      if (typeof stderr === "string") {
        if (stderr.indexOf("No such image") >= 0) {
          err = `No such image: ${this.targetImage}`;
        } else {
          err = error;
        }
      }
    }

    if (callback) {
      try {
        return callback(err, tmpobj.name);
      } finally {
        // We don't need the file anymore and could manually call the removeCallback
        tmpobj.removeCallback();
      }
    }
    // if we didn't pass the keep option the file will be deleted on exit
  }

  /**
   * Analyze and return key files textual content and MD5 sum.
   * @param imageTarPath path to image file saved in tar format
   * @param txtPatterns list of plain text key files paths patterns to extract and return as strings
   * @param md5Patterns list of binary key files paths patterns to extract and return their MD5 sum
   * @returns key files textual content and MD5 sum
   */
  public async analyze(
    imageTarPath: string,
    txtPatterns: string[],
    md5Patterns: string[] = [],
  ) {
    const result: ExtractedKeyFiles = { txt: {}, md5: {} };

    const extracted: ExtractedImage = await extractImageKeyFiles(
      imageTarPath,
      txtPatterns,
      md5Patterns,
    );

    const manifest = JSON.parse(extracted.manifest);
    const layersNames: string[] = manifest[0].Layers;

    if (extracted.layers) {
      // reverse layer order from last to first
      for (const layerName of layersNames.reverse()) {
        // files found for this layer
        if (layerName in extracted.layers) {
          // go over plain text files found in this layer
          for (const filename of Object.keys(extracted.layers[layerName].txt)) {
            // file was not found in previous layer
            if (!Reflect.has(result.txt, filename)) {
              result.txt[filename] = extracted.layers[layerName].txt[filename];
            }
          }
          // go over MD5 sums found in this layer
          for (const filename of Object.keys(extracted.layers[layerName].md5)) {
            // file was not found in previous layer
            if (!Reflect.has(result.md5, filename)) {
              result.md5[filename] = extracted.layers[layerName].md5[filename];
            }
          }
        }
      }
    }
    return result;
  }

  /**
   * Saves the docker image as a TAR file to a temporary location and extract
   * the specified files from it.
   * @param txtPatterns list of plain text key files paths patterns to extract and return as strings
   * @param md5Patterns list of binary key files paths patterns to extract and return their MD5 sum
   * @return list of plain text files and list of MD5 sums
   */
  public async extract(
    txtPatterns: string[],
    md5Patterns: string[] = [],
  ): Promise<ExtractedKeyFiles> {
    return this.save(async (err, imageTarPath) => {
      if (err) {
        throw err;
      }
      return await this.analyze(imageTarPath, txtPatterns, md5Patterns);
    });
  }
}
