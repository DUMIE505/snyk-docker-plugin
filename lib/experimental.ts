import { PluginResponse } from "./types";

export async function experimentalAnalysis(
  options: any,
): Promise<PluginResponse> {
  // assume Distroless scanning
  return distroless(options);
}

// experimental flow expected to be merged with the static analysis when ready
export async function distroless(options: any): Promise<PluginResponse> {
  // assumption #1: the image is present in the local Docker daemon
  // assumption #2: the `docker` binary is available locally
  throw new Error("not implemented");
}
