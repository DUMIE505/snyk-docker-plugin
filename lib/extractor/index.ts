import { extractDockerArchiveLayers } from "./layer";
import { ExtractAction, ExtractedLayers } from "./types";

/**
 * Given a path on the file system to a docker-archive, open it up to inspect the layers
 * and look for specific files. File content can be transformed with a custom callback function if needed.
 * @param fileSystemPath Path to an existing docker-archive.
 * @param extractActions This denotes a file pattern to look for and how to transform the file if it is found.
 * By default the file is returned raw if no processing is desired.
 */
async function getDockerArchiveLayers(
  fileSystemPath: string,
  extractActions: ExtractAction[],
): Promise<ExtractedLayers> {
  const layers = await extractDockerArchiveLayers(
    fileSystemPath,
    extractActions,
  );

  if (!layers) {
    return {};
  }

  const result: ExtractedLayers = {};

  // TODO: This removes the information about the layer name, maybe we would need it in the future?

  for (const layer of layers) {
    // go over extracted files products found in this layer
    for (const filename of Object.keys(layer)) {
      // file was not found
      if (!Reflect.has(result, filename)) {
        result[filename] = layer[filename];
      }
    }
  }

  return result;
}

export { extractDockerArchiveLayers, getDockerArchiveLayers };
