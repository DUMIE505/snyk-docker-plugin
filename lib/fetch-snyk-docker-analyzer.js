const os = require('os');
const fs = require('fs');
const fsExtra = require('fs-extra');
const path = require('path');
const request = require('request');
const ciInfo = require('ci-info');
const ProgressBar = require('progress');
const pkgInfo = require('../package.json')

module.exports = fetch

const version = pkgInfo['snyk-docker-analyzer']['version'];

function getBinaryName() {
  const arch = os.arch();
  if (arch !== 'x64') {
    throw new Error(`Unsupported arch ${arch} - only amd64 is supported`);
  }
  const suffix = '';
  var platform = os.platform();
  if (platform === 'win32') {
    platform = 'windows';
    suffix = '.exe';
  }
  return `snyk-docker-analyzer-${platform}-amd64${suffix}`;
}

function getBinaryLocalPath() {
  const name = getBinaryName();
  return path.join(__dirname, '../bin/', version, name)
}

function fetch() {
  return new Promise((resolve, reject) => {
    try {
      const localPath = getBinaryLocalPath()
      if (fs.existsSync(localPath)) {
        return resolve(localPath);
      }
      fsExtra.ensureDirSync(path.dirname(localPath));
      const downloadUrl =
        `https://s3.amazonaws.com/snyk-docker-analyzer-releases/${version}/${getBinaryName()}`; // jscs:ignore maximumLineLength

      var bar;
      const fsStream = fs.createWriteStream(localPath + '.part');
      const req = request(downloadUrl);
      req
        .on('response', function (res) {
          if (res.statusCode >= 400) {
            var err = new Error(
              'Bad HTTP response for snyk-docker-analyzer download');
            err.statusCode = res.statusCode;
            fsStream.destroy();
            reject(err);
            req.abort();
            return;
          }

          if (ciInfo.isCI) {
            console.log(`downloading ${getBinaryName()} ...`);
          } else {
            const total = parseInt(res.headers['content-length'], 10);
            bar = new ProgressBar(`downloading ${getBinaryName()} [:bar] :rate/Kbps :percent :etas remaining`, { // jscs:ignore maximumLineLength
              complete: '=',
              incomplete: '.',
              width: 20,
              total: total / 1000,
            });
          }
        })
        .on('data', function (chunk) {
          if (bar) {
            bar.tick(chunk.length / 1000);
          }
        })
        .on('error', function (err) {
          reject(err);
        })
        .pipe(fsStream)
        .on('error', function (err) {
          fsStream.destroy();
          reject(err);
        })
        .on('finish', function () {
          fs.renameSync(localPath + '.part', localPath);
          const CHMOD_WITH_EXEC = 0755;
          fs.chmodSync(localPath, CHMOD_WITH_EXEC);
          resolve(localPath);
        })
    } catch (err) {
      reject(err);
    }
  })
}
