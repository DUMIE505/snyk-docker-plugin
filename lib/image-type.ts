import { ImageType } from "./types";

export function getImageType(targetImage: string): ImageType {
  return ImageType.Identifier;
}
