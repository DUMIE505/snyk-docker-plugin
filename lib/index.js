var fs = require('fs');
var path = require('path');
var subProcess = require('./sub-process');
var fetchSnykDockerAnalyzer = require('./fetch-snyk-docker-analyzer');

module.exports = {
  inspect: inspect,
};

function inspect(root, targetFile) {
  const targetImage = root;
  return fetchSnykDockerAnalyzer()
  .then(function (analyzerBinaryPath) {
    return Promise.all([
      getMetaData(),
      getDependencies(analyzerBinaryPath, targetImage),
    ])
    .then(function (result) {
      return {
        plugin: result[0],
        package: result[1],
      };
    });
  })
}

function getMetaData() {
  return subProcess.execute('docker', ['version'])
  .then(function (output) {
    var runtime;
    versionMatch = /Version:\s+(.*)\n/.exec(output);
    if (versionMatch) {
      runtime = 'docker ' + versionMatch[1]
    }

    return {
      name: 'snyk-docker-plugin',
      runtime: runtime,
    };
  });
}

function getDependencies(analyzerBinaryPath, targetImage) {
  console.log(`analyzing ${targetImage} Docker image`);
  return subProcess.execute(
    analyzerBinaryPath,
    buildArgs(targetImage)
  )
  .then(function (output) {
    scanResults = JSON.parse(output);
    return Promise.resolve(convertDependecies(targetImage, scanResults));
  })
  .catch(function (error) {
    if (typeof error === 'string') {
      if (error.indexOf('command not found') !== -1) {
        throw new Error('Snyk docker CLI wasn\'t found')
      }
      errorMsg = error;
      errorMatch = /msg="(.*)"/g.exec(errorMsg)
      if (errorMatch) {
        errorMsg = errorMatch[1]
      }
      throw new Error(errorMsg);
    }

    throw error;
  })
}

function convertDependecies(targetImage, scanResults) {
  targetSplit = targetImage.split(':');
  imageName = targetSplit[0];
  imageVersion = targetSplit[1];
  imageVersion = (imageVersion ? imageVersion : 'lateset');

  root = {};
  root.name = imageName;
  root.version = imageVersion;
  root.from = [imageName + '@' + imageVersion];

  var scanResult = scanResults.filter(function (res) {
    return res.Analysis && res.Analysis.length > 0;
  })[0];

  var pkgType;
  switch (scanResult.AnalyzeType) {
    case 'Apt': {
      pkgType = 'deb';
      break;
    }
    default: {
      pkgType = scanResult.AnalyzeType.toLowerCase()
    }
  }
  root.packageFormatVersion = pkgType + ':0.0.1';

  pkgs = scanResult['Analysis']

  root.dependencies = pkgs.reduce(function (acc, pkg) {
    name = pkg['Name'];
    version = pkg['Version'];

    acc[name] = {
      name: name,
      version: version,
      dependencies: {},
      from: [
        root['from'][0],
        name + '@' + version,
      ],
    }
    return acc;
  }, {});

  return root;
}

function buildArgs(targetImage) {
  var args = ['analyze', targetImage];
  return args;
}

function pathToPosix(fpath) {
  var parts = fpath.split(path.sep);
  return parts.join(path.posix.sep);
}
