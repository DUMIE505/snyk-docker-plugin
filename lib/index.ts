import * as dockerFile from "./docker-file";
import { experimentalAnalysis } from "./experimental";
import * as staticUtil from "./static";
import {
  AppDepsScanResult,
  OsDepsScanResult,
  ScanOptions,
  ScanResult,
} from "./types";

export { scan, dockerFile, OsDepsScanResult, AppDepsScanResult, ScanResult };

async function scan(
  root: string,
  targetFile?: string,
  options?: Partial<ScanOptions>,
): Promise<ScanResult[]> {
  const targetImage = root;

  const dockerfileAnalysis = await dockerFile.readDockerfileAndAnalyse(
    targetFile,
  );

  if (options && options.experimental) {
    return await experimentalAnalysis(targetImage, dockerfileAnalysis, options);
  }

  return await staticUtil.analyzeStatically(
    targetImage,
    dockerfileAnalysis,
    options,
  );
}

setImmediate(async () => {
  const result = await scan("snyk/kubernetes-monitor:1.32.2", "Dockerfile", {
    experimental: true,
    appScan: true,
    "app-vulns": true,
  });
  // tslint:disable-next-line: no-console
  console.log(JSON.stringify(result));
});
