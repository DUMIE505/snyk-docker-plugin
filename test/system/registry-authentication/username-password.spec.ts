import { EOL } from "os";
import { scan } from "../../../lib/index";

describe("username and password authentication", () => {
  it("should correctly authenticate to the container registry when username and password are provided", async () => {
    const pluginResult = await scan({
      path: process.env.DOCKER_HUB_PRIVATE_IMAGE,
      username: process.env.DOCKER_HUB_USERNAME,
      password: process.env.DOCKER_HUB_PASSWORD,
    });

    expect(pluginResult).toMatchSnapshot();
  });
});

describe("handles bad input being provided", () => {
  it("should reject when provided with bad credentials", async () => {
    await expect(() =>
      scan({
        path: process.env.DOCKER_HUB_PRIVATE_IMAGE,
        username: "foo",
        password: "bar",
      }),
    ).rejects.toEqual(
      Error(`{"details":"incorrect username or password"}${EOL}`),
    );
  });
});
