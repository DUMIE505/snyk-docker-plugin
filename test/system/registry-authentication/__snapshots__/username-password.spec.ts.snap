// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`username and password authentication should correctly authenticate to the container registry when username and password are provided 1`] = `
Object {
  "scanResults": Array [
    Object {
      "facts": Array [
        Object {
          "data": Object {
            "graph": Object {
              "nodes": Array [
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "alpine-baselayout/alpine-baselayout@3.1.0-r0",
                    },
                    Object {
                      "nodeId": "alpine-keys/alpine-keys@2.1-r1",
                    },
                    Object {
                      "nodeId": "apk-tools/apk-tools@2.10.1-r0",
                    },
                    Object {
                      "nodeId": "busybox/busybox@1.28.4-r3|2",
                    },
                    Object {
                      "nodeId": "busybox/ssl_client@1.28.4-r3",
                    },
                    Object {
                      "nodeId": "gcc/libgcc@6.4.0-r9",
                    },
                    Object {
                      "nodeId": "gcc/libstdc++@6.4.0-r9|2",
                    },
                    Object {
                      "nodeId": "libc-dev/libc-utils@0.7.1-r0",
                    },
                    Object {
                      "nodeId": "libressl/libressl2.7-libcrypto@2.7.4-r0|2",
                    },
                    Object {
                      "nodeId": "libressl/libressl2.7-libssl@2.7.4-r0|2",
                    },
                    Object {
                      "nodeId": "libressl/libressl2.7-libtls@2.7.4-r0|2",
                    },
                    Object {
                      "nodeId": "musl/musl@1.1.19-r10",
                    },
                    Object {
                      "nodeId": "musl/musl-utils@1.1.19-r10|2",
                    },
                    Object {
                      "nodeId": "pax-utils/scanelf@1.2.3-r0|2",
                    },
                    Object {
                      "nodeId": "zlib/zlib@1.2.11-r1|2",
                    },
                  ],
                  "nodeId": "root-node",
                  "pkgId": "docker-image|snykgoof/dockerhub-goof@alpine",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "busybox/busybox@1.28.4-r3|1",
                  "pkgId": "busybox/busybox@1.28.4-r3",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "musl/musl@1.1.19-r10",
                    },
                  ],
                  "nodeId": "busybox/busybox@1.28.4-r3|2",
                  "pkgId": "busybox/busybox@1.28.4-r3",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "musl/musl@1.1.19-r10",
                  "pkgId": "musl/musl@1.1.19-r10",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "busybox/busybox@1.28.4-r3|1",
                    },
                    Object {
                      "nodeId": "musl/musl@1.1.19-r10",
                    },
                  ],
                  "nodeId": "alpine-baselayout/alpine-baselayout@3.1.0-r0",
                  "pkgId": "alpine-baselayout/alpine-baselayout@3.1.0-r0",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "alpine-keys/alpine-keys@2.1-r1",
                  "pkgId": "alpine-keys/alpine-keys@2.1-r1",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libressl/libressl2.7-libcrypto@2.7.4-r0|1",
                  "pkgId": "libressl/libressl2.7-libcrypto@2.7.4-r0",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "musl/musl@1.1.19-r10",
                    },
                  ],
                  "nodeId": "libressl/libressl2.7-libcrypto@2.7.4-r0|2",
                  "pkgId": "libressl/libressl2.7-libcrypto@2.7.4-r0",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libressl/libressl2.7-libssl@2.7.4-r0|1",
                  "pkgId": "libressl/libressl2.7-libssl@2.7.4-r0",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "libressl/libressl2.7-libcrypto@2.7.4-r0|1",
                    },
                    Object {
                      "nodeId": "musl/musl@1.1.19-r10",
                    },
                  ],
                  "nodeId": "libressl/libressl2.7-libssl@2.7.4-r0|2",
                  "pkgId": "libressl/libressl2.7-libssl@2.7.4-r0",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "zlib/zlib@1.2.11-r1|1",
                  "pkgId": "zlib/zlib@1.2.11-r1",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "musl/musl@1.1.19-r10",
                    },
                  ],
                  "nodeId": "zlib/zlib@1.2.11-r1|2",
                  "pkgId": "zlib/zlib@1.2.11-r1",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "libressl/libressl2.7-libcrypto@2.7.4-r0|1",
                    },
                    Object {
                      "nodeId": "libressl/libressl2.7-libssl@2.7.4-r0|1",
                    },
                    Object {
                      "nodeId": "musl/musl@1.1.19-r10",
                    },
                    Object {
                      "nodeId": "zlib/zlib@1.2.11-r1|1",
                    },
                  ],
                  "nodeId": "apk-tools/apk-tools@2.10.1-r0",
                  "pkgId": "apk-tools/apk-tools@2.10.1-r0",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libressl/libressl2.7-libtls@2.7.4-r0|1",
                  "pkgId": "libressl/libressl2.7-libtls@2.7.4-r0",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "libressl/libressl2.7-libcrypto@2.7.4-r0|1",
                    },
                    Object {
                      "nodeId": "libressl/libressl2.7-libssl@2.7.4-r0|1",
                    },
                    Object {
                      "nodeId": "musl/musl@1.1.19-r10",
                    },
                  ],
                  "nodeId": "libressl/libressl2.7-libtls@2.7.4-r0|2",
                  "pkgId": "libressl/libressl2.7-libtls@2.7.4-r0",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "libressl/libressl2.7-libtls@2.7.4-r0|1",
                    },
                    Object {
                      "nodeId": "musl/musl@1.1.19-r10",
                    },
                  ],
                  "nodeId": "busybox/ssl_client@1.28.4-r3",
                  "pkgId": "busybox/ssl_client@1.28.4-r3",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "musl/musl@1.1.19-r10",
                    },
                  ],
                  "nodeId": "gcc/libstdc++@6.4.0-r9|1",
                  "pkgId": "gcc/libstdc++@6.4.0-r9",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "gcc/libstdc++@6.4.0-r9|2",
                  "pkgId": "gcc/libstdc++@6.4.0-r9",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "gcc/libstdc++@6.4.0-r9|1",
                    },
                    Object {
                      "nodeId": "musl/musl@1.1.19-r10",
                    },
                  ],
                  "nodeId": "gcc/libgcc@6.4.0-r9",
                  "pkgId": "gcc/libgcc@6.4.0-r9",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "musl/musl-utils@1.1.19-r10|1",
                  "pkgId": "musl/musl-utils@1.1.19-r10",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "musl/musl@1.1.19-r10",
                    },
                    Object {
                      "nodeId": "pax-utils/scanelf@1.2.3-r0|1",
                    },
                  ],
                  "nodeId": "musl/musl-utils@1.1.19-r10|2",
                  "pkgId": "musl/musl-utils@1.1.19-r10",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "musl/musl-utils@1.1.19-r10|1",
                    },
                  ],
                  "nodeId": "libc-dev/libc-utils@0.7.1-r0",
                  "pkgId": "libc-dev/libc-utils@0.7.1-r0",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "pax-utils/scanelf@1.2.3-r0|1",
                  "pkgId": "pax-utils/scanelf@1.2.3-r0",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "musl/musl@1.1.19-r10",
                    },
                  ],
                  "nodeId": "pax-utils/scanelf@1.2.3-r0|2",
                  "pkgId": "pax-utils/scanelf@1.2.3-r0",
                },
              ],
              "rootNodeId": "root-node",
            },
            "pkgManager": Object {
              "name": "apk",
              "repositories": Array [
                Object {
                  "alias": "alpine:3.8.2",
                },
              ],
            },
            "pkgs": Array [
              Object {
                "id": "docker-image|snykgoof/dockerhub-goof@alpine",
                "info": Object {
                  "name": "docker-image|snykgoof/dockerhub-goof",
                  "version": "alpine",
                },
              },
              Object {
                "id": "busybox/busybox@1.28.4-r3",
                "info": Object {
                  "name": "busybox/busybox",
                  "version": "1.28.4-r3",
                },
              },
              Object {
                "id": "musl/musl@1.1.19-r10",
                "info": Object {
                  "name": "musl/musl",
                  "version": "1.1.19-r10",
                },
              },
              Object {
                "id": "alpine-baselayout/alpine-baselayout@3.1.0-r0",
                "info": Object {
                  "name": "alpine-baselayout/alpine-baselayout",
                  "version": "3.1.0-r0",
                },
              },
              Object {
                "id": "alpine-keys/alpine-keys@2.1-r1",
                "info": Object {
                  "name": "alpine-keys/alpine-keys",
                  "version": "2.1-r1",
                },
              },
              Object {
                "id": "libressl/libressl2.7-libcrypto@2.7.4-r0",
                "info": Object {
                  "name": "libressl/libressl2.7-libcrypto",
                  "version": "2.7.4-r0",
                },
              },
              Object {
                "id": "libressl/libressl2.7-libssl@2.7.4-r0",
                "info": Object {
                  "name": "libressl/libressl2.7-libssl",
                  "version": "2.7.4-r0",
                },
              },
              Object {
                "id": "zlib/zlib@1.2.11-r1",
                "info": Object {
                  "name": "zlib/zlib",
                  "version": "1.2.11-r1",
                },
              },
              Object {
                "id": "apk-tools/apk-tools@2.10.1-r0",
                "info": Object {
                  "name": "apk-tools/apk-tools",
                  "version": "2.10.1-r0",
                },
              },
              Object {
                "id": "libressl/libressl2.7-libtls@2.7.4-r0",
                "info": Object {
                  "name": "libressl/libressl2.7-libtls",
                  "version": "2.7.4-r0",
                },
              },
              Object {
                "id": "busybox/ssl_client@1.28.4-r3",
                "info": Object {
                  "name": "busybox/ssl_client",
                  "version": "1.28.4-r3",
                },
              },
              Object {
                "id": "gcc/libstdc++@6.4.0-r9",
                "info": Object {
                  "name": "gcc/libstdc++",
                  "version": "6.4.0-r9",
                },
              },
              Object {
                "id": "gcc/libgcc@6.4.0-r9",
                "info": Object {
                  "name": "gcc/libgcc",
                  "version": "6.4.0-r9",
                },
              },
              Object {
                "id": "musl/musl-utils@1.1.19-r10",
                "info": Object {
                  "name": "musl/musl-utils",
                  "version": "1.1.19-r10",
                },
              },
              Object {
                "id": "libc-dev/libc-utils@0.7.1-r0",
                "info": Object {
                  "name": "libc-dev/libc-utils",
                  "version": "0.7.1-r0",
                },
              },
              Object {
                "id": "pax-utils/scanelf@1.2.3-r0",
                "info": Object {
                  "name": "pax-utils/scanelf",
                  "version": "1.2.3-r0",
                },
              },
            ],
            "schemaVersion": "1.2.0",
          },
          "type": "depGraph",
        },
        Object {
          "data": Array [
            "0fc4cab19958b98d637fefa8a81af2fb42aec5a9dcad2f636aad4a58a665181c",
          ],
          "type": "keyBinariesHashes",
        },
        Object {
          "data": "ebbf98230a821f1795b10c4df40ce81b1905e47d625cb9b4b0c7c617acb85e53",
          "type": "imageId",
        },
        Object {
          "data": Array [
            "169185f82c45a6eb72e0ca4ee66152626e7ace92a0cbc53624fb46d0a553f0bd/layer.tar",
            "80e1b3e484f00bce3ac164eb0d744be783f154f356f79491f2a0236a1a6ecae8/layer.tar",
            "93e33f1be740ed7d552b1f02a7d766f8a51704e9b889e8c70e293b2b3b8d74a5/layer.tar",
          ],
          "type": "imageLayers",
        },
        Object {
          "data": Array [
            "sha256:767f936afb51c8a3ad9a96592a4be092048bb70f2ca410028a0b3f64b826acbb",
            "sha256:52506d3645613bb715039c51c2076e89b1a370042b044bf8023968180cc71277",
            "sha256:bc39a721343bd8a32be9237b976929718f7b0e66bdc8834d8f8cb79c80d7d08f",
          ],
          "type": "rootFs",
        },
        Object {
          "data": "Alpine Linux v3.8",
          "type": "imageOsReleasePrettyName",
        },
      ],
      "identity": Object {
        "args": Object {
          "platform": "linux/amd64",
        },
        "type": "apk",
      },
      "target": Object {
        "image": "docker-image|snykgoof/dockerhub-goof",
      },
    },
  ],
}
`;
